tools-local/Microsoft.ML.InternalCodeAnalyzer/BestFriendAnalyzer.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.ML.InternalCodeAnalyzer
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class BestFriendAnalyzer : DiagnosticAnalyzer
+    {
+        private const string Category = "Access";
+        internal const string DiagnosticId = "MSML_NoBestFriendInternal";
+
+        private const string Title = "Cross-assembly internal access requires.";
+        private const string Format = "Access of '{0}' is a cross assembly internal " +
+            "reference, and the declaring assembly wants these accesses to be on something " +
+            "with the attribute " + AttributeName + ".";
+        private const string Description =
+            "The ML.NET .";
+
+        private static DiagnosticDescriptor Rule =
+            new DiagnosticDescriptor(DiagnosticId, Title, Format, Category,
+                DiagnosticSeverity.Warning, isEnabledByDefault: true, description: Description);
+
+        private const string AttributeName = "Microsoft.ML.BestFriendAttribute";
+        private const string AssemblyAttributeName = "Microsoft.ML.WantsToBeBestFriendsAttribute";
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterSemanticModelAction(Analyze);
+        }
+
+        private void Analyze(SemanticModelAnalysisContext context)
+        {
+            var model = context.SemanticModel;
+            var comp = model.Compilation;
+
+            // Get the symbols of the key types we are analyzing. If we can't find any of them there is
+            // no point in going further.
+            var attrType = comp.GetTypeByMetadataName(AttributeName);
```

Comment:
Would it make sense to rename `attrType` to `bestFriendAttributeType`? Similar for `assemblyAttrType` => `wantsToBeBestFriendsAttributeType`?

It gets a little confusing below when you see `attrType`, and you kind of think "which attribute are we talking about here?"

Reply:
Yeah, that's what I was going for with the `attr` vs. `assemblyAttr` to track the two, but certainly it doesn't hurt to be more explicit.